### PR TITLE
Draft: provide the possibility to extend support for Draft Edit in every object

### PR DESCRIPTION
**Pull request by carlopav**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by carlopav** [comment]
ref https://forum.freecadweb.org/viewtopic.php?f=23&t=44965&p=423588#p423588

Yiad2JyPiv

**Comment by Vanuan** [comment]
It looks like Arch and Draft are tightly coupled. I'd rather find a way to extract Arch functionality from Draft rather than modifying it directly.

**Comment by carlopav** [comment]
> 
> 
> It looks like Arch and Draft are tightly coupled. I'd rather find a way to extract Arch functionality from Draft rather than modifying it directly.

Well, I think the point is that Draft if not only a wb for 2d drafting, but it's the seed for direct modelling in FreeCAD (https://wiki.freecadweb.org/Direct_modeling_tools).
So whenever you need Snapper or Edit you have to rely on it, or you can reimplement editing workflows like Chris_G did with Curves WB. What i'd like is that such tools in the future should be part of the core Gui module, offering those capabilities to every other workbench, but that's just a draftsman point of view.

Another thing i was thinking is that we can provide a `.get_editpoints` and `.update_from_editpoints` methods right in the objects, so what Draft_Edit does is just to ask for the local position of points and return the modified point. 
But this indeed makes more complicated to handle the preview and the context menu i guess... what do you think?

**Comment by Vanuan** [comment]
> Another thing i was thinking is that we can provide a .get_editpoints and .update_from_editpoints methods right in the objects, so what Draft_Edit does is just to ask for the points and return the modified points, allowing every other computation to the object itself.

Pretty much what I was thinking.

Whenever I see `if type(obj) = 'Type'` I think of a way to embed this functionality into an object, so that I can create more object types without touching this module ([open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle)).

**Comment by carlopav** [comment]
> Pretty much what I was thinking.
> 
> Whenever I see `if type(obj) = 'Type'` I think of a way to embed this functionality into an object, so that I can create more object types without touching this module ([open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle)).

Ok, it seems good, expecially because those objects are new and we do not have backward compatibility problems with them. So we can keep the editing code for objects that we can't modify (Part, Sketcher, Draft objects) and add a hook for other objects.

Perhaps the object viewprovider can contain a hook for coin code of the preview?

@yorikvanhavre wait before merging, I'd like to change that :)

**Comment by carlopav** [comment]
@Vanuan something like this?

**Comment by Vanuan** [comment]
@carlopav  That looks better. Though, I would put these methods to a separate class. E.g.:

```
def get_obj_gui_tools(obj):
    try:
        obj_gui_tools = obj.gui_tools
    except AttributeError:
        try:
            obj_gui_tools = obj.Proxy.gui_tools
        except AttributeError:
           obj_gui_tools = None
    return obj_gui_tools

...
# interface
class GuiTools:
    def __init__(self, obj):
        pass
    def get_edit_point_context_menu(self):
        pass
...

# common code for all objects
obj_gui_tools = get_obj_gui_tools(obj)
if obj_gui_tools:
    obj_gui_tools.get_edit_point_context_menu(...)
...

# object-specific implementation
class WallGuiTools(GuiTools):
...

class Wall():
   def __init__(self):
       self.obj_gui_tools = WallGuiTools()

```

An alternative, where tools are decoupled from the Wall class:

```
# common code
gui_tools_repository = GuiToolsRepository()
class GuiToolsRepository():
    def __init__(self):
         self.obj_gui_tools = {}

    def get(obj):
         return self.obj_gui_tools[type(obj)]

    def add(self, type, gui_tools):
        self.obj_gui_tools[type] = gui_tools

obj_gui_tools = gui_tools_repository.get(obj)
if obj_gui_tools:
    obj_gui_tools.get_edit_point_context_menu(...)

# object-specific code, e.g. wall.py
gui_tools_repository.add('Wall', WallGuiTools())
```

Looks complicated, but IMHO is more flexible in the long run

**Comment by carlopav** [comment]
Interesting, let me understand, the first option allows you to define a custom GuiTools object for every Object, right? that contains all the necessary methods. Could we also multiple inherit WallGuiTools from Wall object or from the ViewProviderWall to have directly access to WallGuiTools's methods from within the wall?

the second option: wow. I was trying to implement it in the gui_edit module, but without success because I did try with functions and not with classes... I'll try that for the objects that will continue to be held into the Draft module.

hope I got correctly what you meant. :)

**Comment by Vanuan** [comment]
> the first option allows you to define a custom GuiTools object for every Object, right

Yeap, since the implementation of every object can be different, you may have to customize tools for each object type. But it doesn't mean you can't reuse code. For example, you may already have GuiTools for vector and vertices if those are the base of a high-level object.

> Could we also multiple inherit WallGuiTools from Wall object or from the ViewProviderWall to have directly access to WallGuiTools's methods from within the wall?

I think multiple inheritance is generally discouraged. Composition is preferable and easier to comprehend.

Here's the domain model as I see it:
- Wall is the UI-independent model of the architectural entity, has a base axis (a vector with first_point and last_point) , an extrusion direction, height and width
- Wall view provider is the 3D-engine specific implementation of GUI view of a wall (pivy.coin nodes), provides mapping of the wall model to a graphical scene, contains display options unrelated to architectural entity (visibility, colors, lines, etc)
- GUI tools represent user intent and manipulations with a model (e.g. move vector points)

Implementing this domain model in a single object is technically possible (using multiple inheritance), but I wouldn't call it a good design. User intent to manipulate an object should be modeled as a separate entity. So that when you add more ways to manipulate the object (e.g. swap the first and the last points of a vector), you only touch the "tools" part of the code. And when you add more ways to model the object (e.g. using 2D curves or profile rather than vector), you only touch the UI-independent model (since you may already have tools to manipulate 2D objects such as curves or vector).

So, we have to compose these 3 domain entities somehow. Since view in FreeCAD is implemented as a `.ViewObject` property, the easiest way to implement tools to manipulate the object is to do the same: attach tools as a property of an object.

I'm still not very familiar with a codebase. Maybe there's already something like this somewhere.

**Comment by Vanuan** [comment]
Here's some picture to think about:

wall model

[uml](https://www.planttext.com/?text=VP6nhW8X44LxleBoMUmFMBpOMMfXj9pWCXfUW24mx9eO_zqHDN7rnGxklNEu4nQ5JUOQl5BisAFU6Uzr88kYgDwRKkxIMMczTJYkzd_OyxoTrva6Pp7V0c8VgmCcycNUHfRLRMutSGeox4r166IMyizCNPxmTqvxw2kf9kWfCfuO0iRw2diDDr36daAgZ32A1HXjaRpMDLaZuY5J0BIE8P6BNEwiLgEpDCRkztFFYJjfzx7Alwb9nwwcXw_KGjh9Hrm1)

**Comment by carlopav** [comment]
Seems feasible, even if I see more like this:

immagine

So let me understand: the WallGuiTools object is instantiated only when requested from Draft Edit through the `obj.gui_tools` attribute?

**Comment by Vanuan** [comment]
Instantiation is debatable. There are many creational patterns. What's required is that gui tools has access to the object(s) they operate on.

You could have singleton (module level functions) and pass an object as a parameter, but you lack the ability to have a generic way to call the tools (you have to specify the type of object).

So, an object level instance of tools (composition) or a dedicated repository (factory method) is more flexible.

**Comment by Vanuan** [comment]
As to when exactly you have to create an instance, this is probably not very important. I don't think that to support headless mode you have to instantiate tools object on demand, as tools IMHO should not depend on Coin3D or Qt. To create 3D preview you could just clone the wall object with its own view provider, so tools wouldn't depend on Qt directly.

**Comment by carlopav** [comment]
> 
> 
> As to when exactly you have to create an instance, this is probably not very important. I don't think that to support headless mode you have to instantiate tools object on demand, as tools IMHO should not depend on Coin3D or Qt. To create 3D preview you could just clone the wall object with its own view provider, so tools wouldn't depend on Qt directly.

Not exactly: current preview are dependent on pivy through Draft.Trackers because they are not simple copy of the object to move around but objects that are recomputed at every mouse movement.

Anyway, it's late tonight and being half drunk it's hard to follow you (also because you seem a professional developer while i'm writing python code from just 2 years, being an architect in the real life). Nevertheless I want to thank you for your support, i'm learning a lot in this conversation. It would be cool to have a chat on jistsi. goodnight!

**Comment by Vanuan** [comment]
> seem a professional developer while i've been writing python code for just 2 years during the weekend, being an architect in the real life

Yep, it's easy for me to produce a software architecture rather than architect a building. So my hope is that while designing software I can learn to design a building (my dream is parametric technical interior design). While you seem to know how to design a building, but have some issues with how to express it in the code.

> It would be cool to have a chat on jitsi

See my invitation in hangouts

**Comment by vocx-fc** [comment]
@carlopav As you know I haven't looked into the edit code a lot, and I haven't looked into Arch with detail, but I tend to agree with vanuan. What is important is to clearly structure the code so that we can replace or extend the behavior by changing a function or class here and there. Much of the work that you and I have done in Draft is constrained with trying to maintain compatibility with older objects, so maybe we need to try new ideas, like what you are doing with the new BIM walls. I think that at some point we will need to break compatibility in order to push Draft and Arch forward, so investigating as many things as possible is not a waste of time.

I think that discussing this in the pull request isn't the best; it's better to do it in the forum.

@vanuan I think you have more or less the right idea. Essentially the Wall object contains the actual information of the object and it's the most important piece. The viewprovider should not contain critical information other than appearance. The Gui Tool should modify the object, which in turn causes an update of the viewprovider.

**Comment by Vanuan** [comment]
@vocx-fc To me, it's much easier to have code-related discussions on GitHub. phpBB interface is much worse. In addition, it lacks the highlighting. If there was a choice I would vote for enabling GitHub issues to discuss specific coding issues and leave the forum entirely to users and out of source workbenches.

**Comment by carlopav** [comment]
@Vanuan i'm implementing your suggestion... works just nicely and it's really tidy. 2 questions:
- where should we put the `gui_tools_repository`, i've started to make it a variable of the Edit object, so `Edit.gui_tools_repository`
- should we register into the repository also the GuiTools that are hosted into the objects (ex the WallGuiTools) or just the ones that are found into the Draft workbench?

**Comment by carlopav** [comment]
@Vanuan still something to fix, but it works quite good... what do you think?

File: src/Mod/Draft/draftguitools/gui_edit.py [hunk 1]
```diff
@@ -1029,76 +1002,61 @@ def update(self, obj, nodeIndex, v):
...
+        else:
+            return False
 
-        elif objectType == "Part" and obj.TypeId == "Part::Cone":
-            edit_part.updatePartCone(obj, nodeIndex, v)
 
-        elif objectType == "Part" and obj.TypeId == "Part::Sphere":
-            edit_part.updatePartSphere(obj, nodeIndex, v)
+    def get_obj_type(self, obj):
+        """ Substitute utils.get_type to provide more specific object support.
```
Comment (vocx-fc) [inline src/Mod/Draft/draftguitools/gui_edit.py:1034]: I don't understand this change. Maybe `utils.get_type` should be augmented?

As it is right now, it seems to me that this method just wraps around the `get_type` function.
```py
def get_obj_type(self, obj):
    return utils.get_type(obj)
```

Maybe there is an idea that you want to implement that I am not seeing at the moment.

Comment (carlopav) [reply src/Mod/Draft/draftguitools/gui_edit.py:1034]: Yes, the point is that utils.get_tipe is really limited. I need the function to return the obj.Proxy.Type or the obj.TypeId if the object do not have a Proxy.
So we can know if a Part object is a Box or a Cylinder.

Comment (vocx-fc) [reply src/Mod/Draft/draftguitools/gui_edit.py:1034]: Mmm. I think we can. Draft does not test many Part objects anyway, these are special cases; and finding these special cases in the code shouldn't take a long time. Draft and Arch mostly test for Draft and Arch objects, which have a `Proxy.Type`. For other cases we could just return the `TypeId`, just like you do here.
```py
        if hasattr(obj, 'TypeId'):
            return obj.TypeId
```

Comment (carlopav) [reply src/Mod/Draft/draftguitools/gui_edit.py:1034]: Ok, perfect so! I'll change get_type.

Comment (vocx-fc) [reply src/Mod/Draft/draftguitools/gui_edit.py]: You'd need to adjust the lines that use `get_type` with the deleted types. I did a quick search in Draft and Arch and this is what I found. These are the types that don't use the `Proxy.Type` nor the base `TypeId`.
* `"Shape"`, used in `make_sketch`
* `"Sketch"`, used in `gui_edit`, and `ArchWall`
* `"Part::Line"`, used in `gui_trimex`, and `gui_edit`
* `"Offset2D"`, not used anywhere
* `"Part"`, used in `gui_edit`, `offset`, and `importIFClegacy`
* `"Annotation"`, used in `importDXF`, `getSVG`, and `move`
* `"Mesh"`, used in `gui_snapper`
* `"Points"`, used in `gui_snapper`
* `"Group"`, used in `importIFClegacy`

So, it doesn't seem that a lot of code needs to be changed. You could also make a small change to keep compatibility.
```py
if hasattr(obj, 'TypeId'):
    if obj.TypeId == "Part::Offset2D":
        return "Offset2D"
    elif obj.TypeId == "App::Annotation":
        return "Annotation"
    ...
    return obj.TypeId
```

Comment (carlopav) [reply src/Mod/Draft/draftguitools/gui_edit.py]: ok, let's do it:

- [ ]     "Shape", used in make_sketch
- [ ]     "Sketch", used in gui_edit, and ArchWall
- [ ]     "Part::Line", used in gui_trimex, and gui_edit
- [ ]     "Offset2D", not used anywhere
- [ ]     "Part", used in gui_edit, offset, and importIFClegacy
- [ ]     "Annotation", used in importDXF, getSVG, and move
- [ ]     "Mesh", used in gui_snapper
- [ ]     "Points", used in gui_snapper
- [ ]     "Group", used in importIFClegacy

Comment (Vanuan) [reply src/Mod/Draft/draftguitools/gui_edit.py]: @vocx-fc Is there a rationale for using `TypeId` rather than `isinstance`? Is it because of C++ bridge or something else?

Comment (vocx-fc) [reply src/Mod/Draft/draftguitools/gui_edit.py]: Well, `TypeId` is a native property that objects have, so I think we can just test for it. With `isinstance` we would need to load the class type to test for it. What is the advantage?

Comment (Vanuan) [reply src/Mod/Draft/draftguitools/gui_edit.py]: @vocx-fc The advantage of `isinstance` is twofold:

- it's a part of Python language and provides strict guarantees. While TypeId is a weak string which is a subject to human error.
- isinstance is supported by the type inference tools:

Screenshot from 2020-08-17 05-58-03

This not only helps developers who use IDE (proper autocomplete support), but also guarantees that developers who don't use IDE make no mistake of accessing non-existing property of an object (e.g. mypy linter).

The type of issue that can be prevented by this is demonstrated here:
https://forum.freecadweb.org/viewtopic.php?p=424088#p424088

Comment (vocx-fc) [reply src/Mod/Draft/draftguitools/gui_edit.py]: Well, in this particular case, not every single object has a specific type. For example, a `Part::Cylinder` is an instance of type `Part.Feature`, but it is *not* an instance of `Part.Cylinder`.

Most classes under `Part` I think are wrappers of the OpenCASCADE types. There is a difference between the class of a "shape", and the class of an "object". Objects are what you actually see in the tree view, but I'm not sure how the internal implementations are made. You'd need to ask Werner.

`Part.Line` is not an object, it is a class that determines a topological shape of a line (`Part::TopoShape`), while `obj.TypeId == 'Part::Line'` tells us this is a specific object; it is a specific `Part::Feature` object with certain properties that makes it behave like a parametric line.
```py
>>> isinstance(App.ActiveDocument.Line, Part.Feature)
True
>>> isinstance(App.ActiveDocument.Line, Part.Line)
False
```

And about loading `Part` anyway, yes, my understanding is that we require loading `Part` in many cases, however, Yorik, the developer of Draft/Arch, has always said that we should avoid loading `Part` unless we really need it because otherwise it makes loading of the workbench slow. So, we should delay as much as possible the use of `Part` classes, unless we really need them. Therefore, to test a line, we can simply test the `obj.TypeId`, and often this is enough, instead of loading and testing the type itself (`Part.Line` or something else).

> Does it have something to do with serialization?

I don't think it does. Maybe it's due to the way the wrapping of C++ classes is done. I'm not entirely sure how this works, it is a mystery to me. You'd have to ask Werner.

Comment (Vanuan) [reply src/Mod/Draft/draftguitools/gui_edit.py]: Ok, I see now. FreeCAD has its own `DocumentObject` type system defined using the `PROPERTY_SOURCE` macro. And properties of `DocumentObject` are defined using the `ADD_PROPERTY_TYPE` macro.

For example, `Part::Line` is defined in `PrimitiveFeature.cpp`, `Part::Primitive` class:

https://github.com/FreeCAD/FreeCAD/blob/f56825c79f11b75648a7a0dcc8a9f52bb4574e61/src/Mod/Part/App/PrimitiveFeature.cpp#L250-L260

The macro calls this static method to associate the `Part::Line` string with the `Part::Primitive` C++ class:
https://github.com/FreeCAD/FreeCAD/blob/f56825c79f11b75648a7a0dcc8a9f52bb4574e61/src/Base/Type.cpp#L142-L153

And then, when you use `Document::addObject`, it calls the instantiation method:

https://github.com/FreeCAD/FreeCAD/blob/f56825c79f11b75648a7a0dcc8a9f52bb4574e61/src/Base/Type.cpp#L91-L103

This implementation means, that if you have an object with `TypeId` that starts with `Modulename::`, `Modulename` has already been imported. Otherwise you wouldn't have the object created in the first place.

Ok. This appears to have everything to do with serialization: to be able to change C++ implementation while keeping the FCStd file's object type names.

So, when you check `if obj.TypeId == 'Part::Line'`, the only thing that is guaranteed that this object has properties `X1, Y1, Z1, X2, Y2, Z2`. It means you still have to use `isinstance` to call the methods of this object. The Python class corresponding to `Part::Line` is `PrimitivePy`. The inheritance chain:

`PrimitivePy > PartFeaturePy > App::GeoFeaturePy > DocumentObjectPy > App::ExtensionContainerPy > App::PropertyContainerPy > Base::PersistencePy > Base::BaseClassPy > Base::PyObjectBase > PyObject`

To sum up:
- if we want to access object data, we may use `TypeId`
- if we want to call object methods, we must use `isinstance`
- not all *Py types are exported, so isinstance can't always be used. This can be fixed
- DocumentObject properties defined with `ADD_PROPERTY_TYPE` are not part of *Py types. This can be fixed as well

Comment (vocx-fc) [reply src/Mod/Draft/draftguitools/gui_edit.py]: Yes, for low level details like these, please ask Werner. I've never tried to understand all the C++ code because for user facing tools it's not very important; the current interfaces that are accessible from Python are enough to create user tools, so I've been content with improving these tools, as well as organizing the Python code so that it is easier to extend by others.

If you think it's helpful, I also recommend you to document your findings in the wiki so that other C++ programmers can get into the code faster. Two resources you should be aware of:
* [FreeCAD_Mod_Dev_Guide](https://github.com/qingfengxia/FreeCAD_Mod_Dev_Guide), written by Qingfeng, the original author of the Cfd workbench. This guide describes some details on how the core is structured. He started writing it for v0.16 and I believe he has updated some parts for v0.19. He intentionally placed it in GitHub so that other people could contribute, but so far I don't think anybody but him has added content.
* Realthunder's [Assembly3 repository](https://github.com/realthunder/FreeCAD_assembly3/wiki). He has added a lot of information of his changes with respect to the main branch. Many of these changes were already merged August 2019 with the `LinkMerge` branch, and possibly more commits will come as soon as 0.19 is released as stable.

File: src/Mod/Arch/importIFClegacy.py [hunk 2]
```diff
@@ -1015,15 +1015,17 @@ def export(exportList,filename):
...
+        elif otype == "Foundation":
             ifctype = "Footing"
-        elif ifctype == "Rebar":
+        elif otype == "Rebar":
             ifctype = "ReinforcingBar"
-        elif ifctype in ["Part","Undefined"]:
+        elif otype == "Undefined":
             ifctype = "BuildingElementProxy"
-            
+        elif otype.split("::")[0] == "Part":
```
Comment (carlopav) [inline src/Mod/Arch/importIFClegacy.py:1024]: I thought to do it like this @vocx-fc . What do you think? The other possibility that came to my mind is to have a function `get_object_workbench(obj)`, but that's so randomly used and would have worked the same way...

Comment (vocx-fc) [reply src/Mod/Arch/importIFClegacy.py:1024]: Looks ugly to me, you can use `startswith("Part::")`.

Comment (carlopav) [reply src/Mod/Arch/importIFClegacy.py:1024]: ok! that looks better

Comment (carlopav) [reply src/Mod/Arch/importIFClegacy.py]: done!

**Comment by carlopav** [comment]
@yorikvanhavre, when you have some free time i consider this pr ready.

**Comment by vocx-fc** [comment]
You should rebase this, because now the unit tests are working, unlike some days ago.

**Comment by carlopav** [comment]
Thanks!
